content/en/blog/2023-02-07-building-a-raft-part-3.md

```diff
@@ -0,0 +1,115 @@
+
+## RequestVote Steps
+
+The `RequestVote` function is how candidates carry out elections in remote replicas. The function expects as input a `VoteRequest`, which contains the candidate's term, it's unique identifier, and some details of it's last log entry.
+
+*Note: Abbreviated for brevity*
+```proto
+message VoteRequest {
+    int32 term = 2;           // The candidate's term.
+    string candidateId = 3;   // ID of candidate requesting a vote.
+    int64 lastLogIndex = 4;   // The index of candidate’s last log entry
+    int32 lastLogTerm = 5;    // The term number of candidate’s last log entry
+}
+```
+
+In return, each voter will respond with a `VoteReply`, which contains their own unique identifier, their perspective of the term, and their vote.
+
+```proto
+message VoteReply {
+    string id = 1;            // ID of the Raft node sending the vote.
+    int32 term = 2;           // Current term number
+    bool voteGranted = 3;     // True if the voter has voted for the candidate
+}
+```
+
+#### 1. Vote "no" if the candidate is behind
+
+Just like in `AppendEntries`, we first need to check if the candidate has an up-to-date term. If the candidate requesting the vote has a term less than the recipient's term, that candidate is not up-to-date with the rest of the cluster, and should be denied the vote. On the other hand if the candidate has a term greater than ours, we should ensure that we are a follower and reset our election timeout since we know there is a node in the cluster more up to date. 
+
+#### 2. Vote "yes" if the candidate is ahead or even (but only vote once!)
+
+Each server in a Raft cluster keeps track of who it has most recently voted for (until the next successful heartbeat). If this `votedFor` field is null, then it hasn’t voted for anyone in the current term and is free to vote for the candidate that sent the request.
+
+Here's the full logic for the `RequestVote` function:
+
+```golang
+func (s *RaftServer) RequestVote(ctx context.Context, in *api.VoteRequest) (out *api.VoteReply, err error) {
+	s.Lock()
+	defer s.Unlock()
+	out = &api.VoteReply{Id: s.id, Term: s.currentTerm, VoteGranted: false}
+	lastLogIndex, lastLogTerm := s.lastLogIndexAndTerm()
+
+	// Check if the incoming request has a higher term than ours, if so then there is a node in the cluster 
+	// more up to date, so we ensure we are a follower and reset our election timeout
+	// Note: It's important that we return this node's current term so that the requester can check if it is 
+	// 		 out of date and if so, do the same (revert to follower and reset it's election timeout)
+	if in.Term > s.currentTerm {
+		fmt.Printf("RequestVote: in.Term > s.currentTerm, reverting to follower\n")
```

Comment:
Should this be a comment rather than a print statement?

Reply:
This is a log statement. I can tell you from experience how important having good logging is while implementing Raft.